File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/object/UaWriteVariableValueTransaction.java
package org.opcfoundation.webserver.service.transactions.object;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.opcfoundation.webserver.addressspace.models.UaObjectType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerWebService;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.service.transactions.base.UaWriteTransaction;
import org.opcfoundation.webserver.types.UaObjectId;
import org.opcfoundation.webserver.types.UaObjectIdentifier;
import org.opcfoundation.webserver.types.UaChildId;
import org.opcfoundation.webserver.types.WriteVariableValue;
import org.opcfoundation.webserver.types.message.WriteVariableValueRequest;
import org.opcfoundation.webserver.types.message.WriteVariableValueResponse;
import org.opcfoundation.webapi.service.types.WriteContext;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public class UaWriteVariableValueTransaction extends UaWriteTransaction {
    private final UaObjectIdentifier objectId;
    private final List<WriteVariableValue> variableValues;
    private final NodeManagerWebService nodeManager;

    public UaWriteVariableValueTransaction(
            WriteContext context,
            UaObjectIdentifier objectId,
            Map<Integer, WriteVariableValue> handleIdsAndVariableValues,
            NodeManagerWebService nodeManager)
    {
        super(context,new ArrayList<>(handleIdsAndVariableValues.keySet()));
        this.objectId = objectId;
        this.variableValues = new ArrayList<>(handleIdsAndVariableValues.values());
        this.nodeManager = nodeManager;
    }

    @Override
    public CompletableFuture<Void> execute()
    {
        try {
            UaObjectType objectType = nodeManager.findObjectType(objectId);
            if (null == objectType) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

            UaObject instanceDeclaration = nodeManager.findInstanceDeclaration(objectId);
            Map<UaChildId, Variant> valuesToWrite = new HashMap<>();

            for (WriteVariableValue item: variableValues)
            {
                if (!item.getValue().statusCode().isGood()) continue;
                valuesToWrite.put(item.getVariableId(), item.getValue().getValue());
            }

            WriteVariableValueRequest request = new WriteVariableValueRequest(
                    new UaObjectId(
                            objectId.getId(),
                            instanceDeclaration),
                    valuesToWrite);

            return objectType.onWriteVariablesValue(request).
                        thenAccept(this::setResults).
                        exceptionally(ex -> buildErrorResponse(ex.getCause()));


        } catch (Exception e) {
            buildErrorResponse(e);
        }

        return CompletableFuture.completedFuture(null);
    }

    private void setResults(WriteVariableValueResponse response) {
        for (WriteVariableValue item: variableValues)
        {
            if (!item.getValue().statusCode().isGood())
            {
                results.add(StatusCode.of(StatusCodes.Bad_NotWritable));
                continue;
            }

            StatusCode result = response.getResults().get(item.getVariableId());
            results.add(Objects.requireNonNullElseGet(result, () -> StatusCode.of(StatusCodes.Bad_NotWritable)));
        }
    }


}

File: opcua-webservice-node/src/server/service/transactions/reactiveobject/UaWriteVariableValueTransaction.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaDataValue, UaError, UaVariant } from "opcua-webapi-ts";
import { WriteContext } from "../../..";
import { NodeManagerReactiveObject } from "../../../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaObjectId, UaObjectIdentifier, WriteVariableValue } from "../../../types";
import { WriteVariableValueRequest, WriteVariableValueResponse } from "../../message";
import { UaWriteTransaction } from "../base/UaWriteTransaction";

export class UaWriteVariableValueTransaction extends UaWriteTransaction {
    private readonly objectId: UaObjectIdentifier;
    private readonly variableValues: Array<WriteVariableValue>;
    private readonly nodeManager: NodeManagerReactiveObject;

    constructor(
        context: WriteContext,
        objectId: UaObjectIdentifier,
        handleIdsAndVariableValues: Map<number, WriteVariableValue>,
        nodeManager: NodeManagerReactiveObject,
    ) {
        super(context, [...handleIdsAndVariableValues.keys()]);
        this.objectId = objectId;
        this.variableValues = [...handleIdsAndVariableValues.values()];
        this.nodeManager = nodeManager;
    }

    override async execute(): Promise<void> {
        try {
            const objectType = this.nodeManager.findObjectType(this.objectId);
            if (objectType === null) {
                throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            }

            const instanceDeclaration = this.nodeManager.findInstanceDeclaration(this.objectId);
            const valuesToWrite = new Map<string, UaVariant>();

            for (const item of this.variableValues) {
                if (item.value.statusCode.isNotGood()) continue;
                valuesToWrite.set(item.variableId.toString(), item.value.value);
            }

            const request = new WriteVariableValueRequest(
                new UaObjectId(this.objectId.id, instanceDeclaration),
                valuesToWrite,
            );

            const response = await objectType.onWriteVariablesValue(request);
            this.setResults(response);
        } catch (error) {
            this._results.length = 0;

            const statusCode = error instanceof UaError
                ? error.statusCode
                : makeUaStatusCode(StatusCodes.BadUnexpectedError);

            for (const _handleId of this.handleIds) {
                this._results.push(statusCode);
            }
        }
    }

    private setResults(response: WriteVariableValueResponse): void {
        this._results.length = 0;

        for (const item of this.variableValues) {
            const result = response.results.get(item.variableId.toString());
            this._results.push(result ?? makeUaStatusCode(StatusCodes.BadNotWritable));
        }
    }
}